compiler/src/iree/compiler/Dialect/Stream/Transforms/AutomaticReferenceCounting.cpp

```diff
@@ -0,0 +1,728 @@
+      untilTimepoint.printAsOperand(llvm::dbgs(), *asmState);
+      llvm::dbgs() << "\n";
+    });
+
+    // Remove any predecessor timepoints to the last timeline use
+    // provided based on the coverage map.
+    auto &set = map[baseResource];
+    set.remove_if([&](Value existingTimepoint) {
+      const bool covers = coverage.covers(existingTimepoint, untilTimepoint);
+      LLVM_DEBUG({
+        if (covers) {
+          llvm::dbgs() << "[arc] removing existing timepoint ";
+          existingTimepoint.printAsOperand(llvm::dbgs(), *asmState);
+          llvm::dbgs() << " as it covers ";
+          untilTimepoint.printAsOperand(llvm::dbgs(), *asmState);
+          llvm::dbgs() << "\n";
+        }
+      });
+      return covers;
+    });
+
+    // Insert the new last use timepoint.
+    set.insert(untilTimepoint);
+  }
+
+  // Inserts a |targetResource| tied to a |sourceResource|. Both resources
+  // are considered equivalent and added to an alias map. This is
+  // conceptually equivalent to:
+  //   consume(sourceResource, timepoint);
+  //   produce(targetResource, timepoint);
+  void tie(Value sourceResource, Value targetResource, Value timepoint) {
+    assert(timepoint);
+
+    // Add an alias from source -> target.
+    // Future references to the target should always return the base
+    // (which may not be source if source itself is an alias).
+    aliasResource(sourceResource, targetResource);
+
+    // Mark the source (or base, if aliased) as in use until the timepoint
+    // has been reached.
+    consume(sourceResource, timepoint);
+  }
+
+  // Inserts an alias of |targetResource| to |sourceResource|.
+  // If |sourceResource| is itself an alias the base resource will be
+  // referenced instead.
+  void aliasResource(Value sourceResource, Value targetResource) {
+    Value baseResource = lookupResource(sourceResource);
+    LLVM_DEBUG({
+      llvm::dbgs() << "[arc] aliasing ";
+      targetResource.printAsOperand(llvm::dbgs(), *asmState);
+      llvm::dbgs() << " to base resource ";
+      baseResource.printAsOperand(llvm::dbgs(), *asmState);
+      llvm::dbgs() << "\n";
+    });
+    aliases.insert({targetResource, baseResource});
+  }
+
+  // Returns the base resource for |resource| if it has been aliased or
+  // |resource| itself.
+  Value lookupResource(Value resource) {
+    auto aliasIt = aliases.find(resource);
+    if (aliasIt != aliases.end()) {
+      return aliasIt->second;
+    }
+    Value baseResource =
+        IREE::Util::TiedOpInterface::findTiedBaseValue(resource);
+    aliases.insert({resource, baseResource});
+    return baseResource;
+  }
+};
+
+// Returns the last defined SSA value in the block in |timepoints|.
+// All timepoints must be in the same block.
+static Value getLastTimepointInBlock(TimepointSet &timepoints) {
+  if (timepoints.empty()) {
+    return nullptr;
+  } else if (timepoints.size() == 1) {
+    return *timepoints.begin();
+  }
+  Value lastTimepoint;
```

Comment:
You could also just use `mlir::computeTopologicalSort` and use the last entry there.

Reply:
unfortunately that only works on ops, while what we care about here is the values (and the values may be block args and have no ops to sort)